In [21]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt

from sklearn.ensemble import RandomForestClassifier as rfc
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import classification_report, accuracy_score , recall_score , precision_score, f1_score

from prepare import prep_titanic , train_val_test
import prepare

seed = 55

## Exercises
### Create a new notebook, random_forests, and work with titanic data to do the following:

In [2]:
titan = prep_titanic()

In [3]:
t_titan, v_titan, test_titan = train_val_test(titan, 'survived', seed=55)

In [4]:
x_train = t_titan.drop(columns=['survived'])
y_train = t_titan.survived
x_val = v_titan.drop(columns=['survived'])
y_val = v_titan.survived

In [5]:
x_train.age.fillna(28, inplace=True)

In [6]:
x_val.age.fillna(28,inplace=True)

In [7]:
x_train = prepare.ret_dum(x_train)
x_val = prepare.ret_dum(x_val)

In [8]:
t_titan.head()

survived  pclass     sex   age  sibsp  parch      fare  embark_town  \
519         0       3    male  32.0      0      0    7.8958  Southampton   
441         0       3    male  20.0      0      0    9.5000  Southampton   
43          1       2  female   3.0      1      2   41.5792    Cherbourg   
341         1       1  female  24.0      3      2  263.0000  Southampton   
664         1       3    male  20.0      1      0    7.9250  Southampton   

     alone  
519      1  
441      1  
43       0  
341      0  
664      0

#### 1. Fit the Random Forest classifier to your training sample and transform (i.e. make predictions on the training sample) setting the random_state accordingly and setting min_samples_leaf = 1 and max_depth = 10.

In [9]:
rf = rfc(min_samples_leaf = 1, max_depth = 10, random_state = seed)

rf.fit(x_train, y_train)

RandomForestClassifier(max_depth=10, random_state=55)

#### 2. Evaluate your results using the model score, confusion matrix, and classification report.

In [10]:
rf.score(x_train, y_train)

0.9598715890850722

In [18]:
rf.score(x_val, y_val)

0.8208955223880597

Terrible accuracy on validate. Must be over fit.

In [17]:
pd.crosstab(y_train,t_pred)

col_0       0    1
survived          
0         381    3
1          22  217

In [19]:
print(classification_report(y_train, t_pred))

              precision    recall  f1-score   support

           0       0.95      0.99      0.97       384
           1       0.99      0.91      0.95       239

    accuracy                           0.96       623
   macro avg       0.97      0.95      0.96       623
weighted avg       0.96      0.96      0.96       623



#### 3. Print and clearly label the following: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

True Positve == 217
False Positive == 3
False Negative == 22
True Negative == 381

In [22]:
recall_score(y_train, t_pred)

0.9079497907949791

In [23]:
accuracy_score(y_train, t_pred)

0.9598715890850722

In [24]:
precision_score(y_train,t_pred)

0.9863636363636363

F1 score is .96

Support = 239

#### 4. Run through steps increasing your min_samples_leaf and decreasing your max_depth.

In [28]:
train_acc = []
val_acc = []
depth = []

for i in range(5, 20):
    rf = rfc(max_depth = i , random_state = seed)
    rf.fit(x_train, y_train)
    
    train_acc.append(rf.score(x_train, y_train))
    val_acc.append(rf.score(x_val,y_val))
    depth.append(rf.score(x_val, y_val))

In [29]:
trees = pd.DataFrame({'max_depth': depth,
                      'train_acc': train_acc,
                      'val_acc': val_acc})
trees

max_depth  train_acc   val_acc
0    0.805970   0.853933  0.805970
1    0.813433   0.892456  0.813433
2    0.835821   0.916533  0.835821
3    0.835821   0.935795  0.835821
4    0.835821   0.951846  0.835821
5    0.820896   0.959872  0.820896
6    0.820896   0.963082  0.820896
7    0.820896   0.967897  0.820896
8    0.820896   0.974318  0.820896
9    0.820896   0.974318  0.820896
10   0.820896   0.980738  0.820896
11   0.820896   0.982343  0.820896
12   0.820896   0.982343  0.820896
13   0.813433   0.982343  0.813433
14   0.820896   0.982343  0.820896

#### What are the differences in the evaluation metrics? Which performs better on your in-sample data? Why?

The deeper I go the better the models perform on my insample data because it is over fitting. 

#### After making a few models, which one has the best performance (or closest metrics) on both train and validate?

The third to the fifth is the best